src/server/kernel/wseat.cpp

```diff
@@ -621,6 +723,99 @@ void WSeat::notifyFrame(WCursor *cursor)
     d->doNotifyFrame();
 }
 
+// deal with touch event form wlr_cursor
+
+void WSeat::notifyTouchDown(WCursor *cursor, WInputDevice *device, int32_t touch_id, uint32_t time_msec)
+{
+    W_D(WSeat);
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    Q_ASSERT(qwDevice);
+    const QPointF &globalPos = cursor->position();
+
+    auto *state = device->getAttachedData<WSeatPrivate::DeviceState>();
+
+    QWindowSystemInterface::TouchPoint *tp = state->point(touch_id);
+    if (Q_UNLIKELY(tp)) {
+        if (Q_LIKELY(tp->state == QEventPoint::Released)) {
+            notifyTouchFrame();
+        }
+        Q_ASSERT_X(state->point(touch_id) == nullptr, "notifyTouchDown", "Inconsistent touch state");
+    }
+
+    QWindowSystemInterface::TouchPoint newTp;
+    newTp.id = touch_id;
+    newTp.state = QEventPoint::Pressed;
+    newTp.area = QRect(0, 0, 8, 8);
+    newTp.area.moveCenter(globalPos);
+    state->m_points.append(newTp);
+}
+
+void WSeat::notifyTouchMotion(WCursor *cursor, WInputDevice *device, int32_t touch_id, uint32_t time_msec) {
+
+    W_DC(WSeat);
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    Q_ASSERT(qwDevice);
+
+    const QPointF &globalPos = cursor->position();
+    auto *state = device->getAttachedData<WSeatPrivate::DeviceState>();
+    QWindowSystemInterface::TouchPoint *tp = state->point(touch_id);
+
+    Q_ASSERT_X(tp, "notifyTouchMotion", "Inconsistent touch state (got 'motion' without 'down')");
+
+    if (Q_LIKELY(tp)) {
```

Comment:
Not need `if` here, because already using `Q_ASSERT_X`

Reply:
The other places too.

Reply:
In release mode,  here will not ASSERT， `if` can prevents some unnecessary crashes